src/lib/operators/projection.cpp

```diff
@@ -97,48 +98,84 @@ std::shared_ptr<const Table> Projection::_on_execute() {
-
-        // Storing the result in output_segments means that the vector may contain both ReferenceSegments and
-        // ValueSegments. We deal with this later.
-        output_segments[column_id] = std::move(output_segment);
-        expression_evaluator_cost += timer.lap();
+      if (expression->type != ExpressionType::PQPColumn) {
+        all_segments_forwarded = false;
+        continue;
       }
+      // Forward input segment if possible
+      const auto& pqp_column_expression = static_cast<const PQPColumnExpression&>(*expression);
+      output_segments[column_id] = input_chunk->get_segment(pqp_column_expression.column_id);
+      column_is_nullable[column_id] = input_table.column_is_nullable(pqp_column_expression.column_id);
     }
+    forwarding_cost += timer.lap();
 
+    // `output_segments_by_chunk` now contains all forwarded segments.
     output_segments_by_chunk[chunk_id] = std::move(output_segments);
+
+    // All columns are forwarded. We do not need to evaluate newly generated columns.
+    if (all_segments_forwarded) continue;
+
+    // Defines the job that performs the evaluation if the columns are newly generated.
+    auto perform_projection_evaluation = [this, chunk_id, &uncorrelated_subquery_results, expression_count,
+                                          &output_segments_by_chunk, &column_is_nullable]() {
+      auto evaluator = ExpressionEvaluator{left_input_table(), chunk_id, uncorrelated_subquery_results};
+
+      for (auto column_id = ColumnID{0}; column_id < expression_count; ++column_id) {
+        const auto& expression = expressions[column_id];
+        if (expression->type != ExpressionType::PQPColumn) {
+          // Newly generated column - the expression needs to be evaluated
+          auto output_segment = evaluator.evaluate_expression_to_segment(*expression);
+          column_is_nullable[column_id] = column_is_nullable[column_id] || output_segment->is_nullable();
+          // Storing the result in output_segments_by_chunk means that the vector for the separate chunks may contain
+          // both ReferenceSegments and ValueSegments. We deal with this later.
+          output_segments_by_chunk[chunk_id][column_id] = std::move(output_segment);
+        }
+      }
+    };
+    // Evaluate the expression immediately if it contains less than `JOB_SPAWN_THRESHOLD` rows, otherwise wrap
+    // it into a task. The upper bound of the chunk size, which defines if it will be executed in parallel or not,
+    // still needs to be re-evaluated over time to find the value which gives the best performance.
+    constexpr auto JOB_SPAWN_THRESHOLD = ChunkOffset{500};
+    if (input_chunk->size() >= JOB_SPAWN_THRESHOLD) {
+      auto job_task = std::make_shared<JobTask>(perform_projection_evaluation);
+      jobs.push_back(job_task);
+    } else {
+      perform_projection_evaluation();
+      expression_evaluator_cost += timer.lap();
+    }
   }
+  // Instruct the scheduler to execute the jobs.
```

Comment:
Not necessary.

Reply:
Assuming that this was the last thing from your side, I took the liberty of removing that line so that we can merge this quickly.